r/R/dataset-scan.R

```diff
@@ -85,9 +86,30 @@ Scanner$create <- function(dataset,
       # To handle mutate() on Table/RecordBatch, we need to collect(as_data_frame=FALSE) now
       dataset <- dplyr::collect(dataset, as_data_frame = FALSE)
     }
+
+    proj <- c(dataset$selected_columns, dataset$temp_columns)
+
+    if (!is.null(projection)) {
+      if (is.character(projection)) {
+        proj <- proj[projection]
+      } else if (is_list_of(projection, "Expression")) {
+        # TODO: need to check and see if there are any Expressions that are simply
+        # field refs in projections, but are richer expressions in proj?
+        proj <- projection
```

Comment:
@nealrichardson do you have any thoughts on this? I've looked into using `projection[[i]]$field_name` to identify any references to fields, and then replace them with what's in `proj[projection[[i]]$field_name]`, which should work for this case, but looks brittle to me (but maybe I'm being paranoid?).

Reply:
I agree it's not great, but if we know that duckdb is just using the character vector version, let's make a todo jira and punt this.

Reply:
If I'm reading[ DuckDB's source](https://github.com/duckdb/duckdb/blob/979a798219262349e8c2c3d52ef203a04d6ee728/tools/rpkg/src/register.cpp#L86) correctly they are only using strings for projection. I'll pull out the expr + make the appropriate todos/jiras. This might get simpler with the exec node stuff you're working on too / we might be able to push this complexity down to that at some point in the future.

After I pull it out, I'll run some test data through duckdb to ensure that we aren't hitting anything funny there either.